QGIS: 2.18.17 -> 3.0.1 plus required libs

coreyoconnor opened the pull request:
I think the wrapper could use some improvements, but it's functional.

###### Motivation for this change

1. update QGIS to 3.0.1

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

coreyoconnor committed:
pythonPackages.OWSLib: init at 0.16.0

jtojnar commented on pkgs/top-level/python-packages.nix:
```diff
@@ -10065,6 +10065,27 @@ in {
     };
   };
 
+  OWSLib = buildPythonPackage rec {
```
Could you move this to a separate file?

coreyoconnor replied:
Yes. There is a mix of inline and separate file in python-packages. Is the standard to move to separate file?

jtojnar replied:
In the past having module expressions in this file was okay; nowadays, we are moving them to separate files.

FRidh replied:
We are moving Python expressions out of `pkgs/top-level/python-packages.nix` into `pkgs/development/python-modules/<module>/default.nix`. 

Please move the expression there, and call it from `pkgs/top-level/python-packages.nix` using `callPackage ../development/python-modules/<package> { };`.

jtojnar commented on pkgs/top-level/python-packages.nix:
```diff
@@ -10065,6 +10065,27 @@ in {
+      sha256 = "0m05225g1sqd2i8r2riaan33953hfni9wjq7n225snhl7klsb5gc";
+    };
+
+    buildInputs = with self; [ dateutil pep8 pillow pyproj pytz pytest pytestcov requests tox ];
```
Some of these should probably go to `checkInputs`.

jtojnar commented on pkgs/top-level/python-packages.nix:
```diff
@@ -10065,6 +10065,27 @@ in {
+    meta = {
+      description = "Client programming with Open Geospatial Consortium (OGC) web service.";
+      license = licenses.bsd3;
+      homepage = "http://geopython.github.io/OWSLib/";
```
URLs need not be quoted.

jtojnar replied:
URLs do not need to be quoted.

jtojnar commented on pkgs/top-level/qscintilla-pyqt5-widgets.patch:
```diff
@@ -0,0 +1,12 @@
+diff -u -r orig/Python/configure.py QScintilla_gpl-2.10.3/Python/configure.py
```
This definitely cannot be in top-level. Should go to `pkgs/development/python-modules/qscintilla/`.

coreyoconnor replied:
whoops! editor goof

jtojnar commented on pkgs/applications/gis/qgis/default.nix:
```diff
@@ -1,42 +1,48 @@
+  # Build processes depend on the built libraries. Extend
+  # LD_LIBRARY_PATH as required.
+  preConfigure = ''
+    export LD_LIBRARY_PATH=`pwd`/build/output/lib:${stdenv.lib.makeLibraryPath [ openssl ]}$LD_LIBRARY_PATH
```
Would not adding `-DCMAKE_SKIP_BUILD_RPATH=OFF` to `cmakeFlags` remove the need for this hack this?

coreyoconnor replied:
No idea. That was from the previous QGIS version. I'll try that option and see.

jtojnar replied:
Apparently the `preConfigure` section is not required any more.

jtojnar commented on pkgs/applications/gis/qgis/default.nix:
```diff
@@ -1,42 +1,48 @@
-  preBuild = ''
-    export LD_LIBRARY_PATH=`pwd`/output/lib:${stdenv.lib.makeLibraryPath [ openssl ]}:$LD_LIBRARY_PATH
-  '';
+  nativeBuildInputs = [ cmake ];
```
Please add `ninja` when using `cmake`, it is usually faster than make.

FRidh commented on pkgs/applications/gis/qgis/default.nix:
```diff
@@ -1,42 +1,48 @@
+  };
 
-  nativeBuildInputs = [ cmake makeWrapper ];
+  pythonBuildInputs = [ python3Packages.qscintilla python3Packages.gdal ] ++
```
This should go into ` let ... in` expression.

FRidh commented on pkgs/applications/gis/qgis/wrapped.nix:
```diff
@@ -0,0 +1,44 @@
+    wrapProgram $out/bin/qgis \
+      --prefix PATH : $program_PATH \
+      --prefix PYTHONPATH : $program_PYTHONPATH \
+      --prefix LD_LIBRARY_PATH : ${stdenv.lib.makeLibraryPath [ openssl ]}
```
this should be patched instead.

jtojnar replied:
This is probably not necessary anymore. `ldd $(nix-build -A qgis)/bin/.qgis-wrapped | grep openssl` shows it is linked correctly.

FRidh commented on pkgs/applications/gis/qgis/wrapped.nix:
```diff
@@ -0,0 +1,44 @@
+
+    wrapProgram $out/bin/qgis \
+      --prefix PATH : $program_PATH \
+      --prefix PYTHONPATH : $program_PYTHONPATH \
```
No prefixing, but setting so PYTHONPATH cannot leak inside.

FRidh commented on pkgs/top-level/all-packages.nix:
```diff
@@ -17335,7 +17335,11 @@ with pkgs;
   qemu-riscv = callPackage ../applications/virtualization/qemu/riscv.nix {};
 
-  qgis = callPackage ../applications/gis/qgis {};
+  qgis-unwrapped = libsForQt5.callPackage ../applications/gis/qgis {
```
Why expose the unwrapped version? Is it used somewhere?

coreyoconnor replied:
Exposing the unwrapped version in these cases is easiest way (IMO) to enable an overlay changing the source of the application. EG:

* https://github.com/coreyoconnor/nix_configs/commit/c37c85afa312e8ba79f9b96cbd4f49add62a9b56

FRidh commented on pkgs/top-level/python-packages.nix:
```diff
@@ -10065,6 +10065,27 @@ in {
+  OWSLib = buildPythonPackage rec {
+    pname = "OWSLib";
+    version = "0.16.0";
+    name = "${pname}-${version}";
```
The `name` attribute is added by `buildPython*` and should therefore be removed.

FRidh reviewed (changes requested):
We should not yet replace 2.8 by 3.0 because 3.0 is not yet feature complete. Instead, add 3.0 next to 2.8.

jtojnar commented:
> We should not yet replace 2.8 by 3.0 because 3.0 is not yet feature complete.

Do you have a source for that? I only found

> This is the first release in the 3.x series. It comes with tons of new features and under-the-hood updates. As such, we do not expect it to be as reliable as the 2.18 LTR just yet.

https://qgis.org/en/site/forusers/visualchangelog30/index.html

FRidh commented:
Its actually the plugins I meant to refer to. Plugins written in 2 are not compatible with 3. As they are a major feature of QGIS we should not be dropping 2 yet. Before 18.09 we can revisit this choice.

coreyoconnor commented:
That explains the mixed messaging I've seen online. The version number and some messaging indicates the QGIS 3 is released. Other messaging (EG: loading screen) states "early adopter".

I just figured it was inconsistency haha. The plugin situation explains that. I'll re-add the 2 derivation.

coreyoconnor committed:
pythonPackages.pyqt5: 5.10 -> 5.10.1

coreyoconnor committed:
pythonPackages.qscintilla: Enable for python 3.

This uses the recommended configuration for python 3 and qt5. The config
used for python 2 and qt 4 is unchanged from before.

coreyoconnor committed:
qgis: add 3.0.1.

coreyoconnor commented:
I re-added QGIS 2.18 as well as resolved the noted issues. Other changes with latest commits:

* Qscintilla is 2.9 not 2.10 as in the original commits. 2.10 has a different library naming convention (`_qtN` suffix) which would change the interface for any existing qscintilla dependents. I don't know if that would matter so I'll defer the upgrade till later. 
* partitioned the python package changes into separate commits.

FRidh commented on pkgs/applications/gis/qgis/3.0.nix:
```diff
@@ -0,0 +1,50 @@
+with lib;
+let
+  pythonBuildInputs = with python3Packages;
+    [ python3Packages.qscintilla gdal jinja2 numpy psycopg2 pygments pyqt5 sip OWSLib six ];
```
repeating `python3Packages`

coreyoconnor replied:
Not a repeat. The "with" does not cause the "qscintilla" to be shadowed. Without the extra python3Packages the "qscintilla" passed is not the one in python3Packages but the C++ library one.

FRidh commented on pkgs/development/python-modules/OWSLib/default.nix:
```diff
@@ -0,0 +1,34 @@
+  buildInputs = [ dateutil pep8 pillow pyproj pytz requests tox ];
+  checkInputs = [ pytest pytestcov ];
+
+  doCheck = false;
```
Why are the tests disabled? Include a comment in the expression explaining why the tests are disabled.

coreyoconnor replied:
No tests run and the build fails because no test run. I figured this was a bug in the python build infrastructure.

FRidh commented on pkgs/development/python-modules/OWSLib/default.nix:
```diff
@@ -0,0 +1,34 @@
+    sha256 = "0m05225g1sqd2i8r2riaan33953hfni9wjq7n225snhl7klsb5gc";
+  };
+
+  buildInputs = [ dateutil pep8 pillow pyproj pytz requests tox ];
```
`cehckInputs`?

coreyoconnor replied:
Already resolved. Not sure why this diff is showing up for you.

FRidh reviewed (changes requested):
Why is the wrapper needed? Can't this be done in the `installPhase` or `fixupPhase`?

Of course it is more convenient when designing the wrappers when not needing to rebuild the whole thing.

coreyoconnor commented:
The wrapper is required to enable extension of the qgis python environment by users without requiring a rebuild of the full application. EG: to add additional deps for plugins for instance.

FRidh commented on pkgs/applications/gis/qgis/wrapped.nix:
```diff
@@ -0,0 +1,46 @@
+let
+  # the python packages that should be included in the PYTHONPATH
+  pythonInputs = with python3Packages;
+    qgis3-unwrapped.pythonBuildInputs ++ [ chardet dateutil pyyaml pytz requests urllib3 ];
```
Why should these few be included in the wrapper, but *not* in the build? Are they needed for specific plugins?

coreyoconnor replied:
Moved to build. The reference errors these resolved only showed up at runtime.
I would expect these are transitive dependencies for packages missing respective propagatedBuildInputs?

FRidh replied:
They are not hard dependencies otherwise the builds of the dependents would fail. Either upstream has forgotten to list their full dependencies, or they require "extras" of certain packages.
https://setuptools.readthedocs.io/en/latest/setuptools.html#declaring-extras-optional-features-with-their-own-dependencies

coreyoconnor commented:
Amended to resolve current CRs.

FRidh commented on pkgs/applications/gis/qgis/wrapped.nix:
```diff
@@ -0,0 +1,43 @@
+      --prefix PATH : $program_PATH \
+      --set PYTHONPATH $program_PYTHONPATH
+
+    # desktop link
```
The following seems irrelevant for the wrapper and thus better placed in the original derivation. Or am I missing something?

coreyoconnor commented:
Times up! no further changes will be submitted.

mpickering commented:
This looks like a great start. I was blocked on this as some dependencies were not building for darwin. If they are resolved I can pick it up.

das-g commented:
I successfully build QGIS 3 from this branch on NixOS 18.03.

How do we proceed? Shall I attempt to resolve the merge conflicts with `master`?

> Times up! no further changes will be submitted.

What does that mean? Was your effort on this time-boxed @coreyoconnor or are you referring to NixOS release 18.03?

das-g commented:
QGIS 3.2 Bonn [has been released](http://blog.qgis.org/2018/06/24/qgis-3-2-bonn-is-released/). Time to revive this PR?

coreyoconnor commented:
Hi! The time I had for this original PR ran out a while ago. I will not have a chance for a few weeks (sadly) to really dig into this again. Anyone should feel free to take the changes thus far and carry on, of course.

I do have a separate branch with these changes rebased against HEAD. Might be able to get this updated accordingly but besides that... a few weeks it is.

coreyoconnor commented:
Clearing old PRs.

das-g commented:
New PR for QGIS 3.4.5 (=the new LTR version and now currently only LTR version): #56486
